Chart md

sparkprime opened the pull request:
I added the new chart.yaml fields, removed fields not supported in the current backend.  Also changed some of the language to reflect this is no longer a proposal but an existing thing.

edit: The intention here is to document existing behavior for users cloning the repo right now.

sparkprime committed:
Major rework of chart_format.md

runseb commented on docs/design/chart_format.md:
```diff
@@ -189,25 +176,14 @@ keywords:
-                extensions:
-                  - extensions/v1beta1
+expander:
+  name: GoTemplates
```
does this work for current helm/charts that are not GoTemplates. In the sense that they are not parameterized ?

sparkprime replied:
The expander is mandatory in the current implementation, so no.  I suppose we could modify the behavior such that if no expander is given in the chart, the chart is simply not expanded.  However this still takes some code to parse it as YAML, handle the errors, and no code exists in manager to do that right now.

sparkprime replied:
Also the string "GoTemplates" may change after Jack integrates that expander.

runseb replied:
fine, but what happens to the current chart that we are talking about merging which don't need expansion.

sparkprime replied:
They won't work for several reasons including that.  Literally the only thing that actually works right now is DM-based charts.  When the go templates expander is integrated, in order to work in the current implementation, they need to be changed to return DM types instead of K8s resources.  I believe then they will work.  That change is one of the items in #520

jackgr replied:
I'll make charts with no expander pass through when adding the pluggable template service.

runseb replied:
ok, I think it is important to keep existing charts running.
Otherwise, why merge them ?

if we can pass through when no expander is specified then fine. but this needs to work in MVP.

so LGTM

sparkprime replied:
+1 to having a semantics for "no expander".  There is probably a good pattern here -- charts with no expander can live at the root and be closed functions.  I will have to change the doc though to reflect that design.

FWIW We want to merge the charts and modify them so they work and are really useful.  E.g. we will probably want to parameterise most of them, and will have to convert them all to return DM types.  So the expander field isn't a big deal in the scheme of things.

jackgr replied:
Not sure what you mean by "live at the root". All content should still be in the `templates` directory, although we might want to rename that directory to `content` as proposed in a comment on an earlier version of this file.

sparkprime replied:
The root of the layout, not the chart.

jackgr replied:
The layout? Do you mean the `templates` directory?

runseb commented on docs/design/chart_format.md:
```diff
@@ -189,25 +176,14 @@ keywords:
-                  - extensions/v1beta1
+expander:
+  name: GoTemplates
+schema: Schema.yaml
```
if the chart is not parameterized, can schema be optional ?

sparkprime replied:
Yeah I forgot to put that schema is optional, will update.

jackgr replied:
Is `Schema.yaml` a valid example of a schema file name, now that we've standardized on JSON schema?

jackgr replied:
We should provide an example of a schema file, at least in snippet form, to help chart authors understand how it's used.

sparkprime commented:
/cc @technosophos

technosophos commented:
I am good with the changes, but will leave the LGTM to @runseb, who had a few questions.

sparkprime commented:
@runseb please take a look at last commit

sparkprime commented:
Fixes #554

jackgr commented on docs/design/chart_format.md:
```diff
@@ -69,16 +66,16 @@ We assume that we are developing a technology that will enjoy widespread use amo
-We have attempted to build a proposal that does not lose any of the functionality of DM, and which loses only Helm functionality that we believe is disposable. But both tools will need to change substantial portions of their chart-handling logic.
-### Non-Goals
-This proposal doesn’t define how either the client or the server must use this data, though we make some non-normative recommendations (e.g. “this data may be displayed to the user” or “a suitable backend will be able to translate this to a fully qualified URL”).
+This design does not lose any of the functionality of DM, and which loses only Helm functionality that we believe is disposable. But both tools have changed substantial portions of their chart-handling logic.
```
Since this is the first time the acronym `DM` is used, the meaning should be spelled out, and the acronym placed in parentheses, like this:
```
Deployment Manager (DM)
```

sparkprime replied:
How about deleting the paragraph entirely?

jackgr replied:
Hmmm. I think it's actually useful, because it provides context for people coming from the previous projects. That said, it's the kind of thing that could live in README.md, not here.

jackgr commented on docs/design/chart_format.md:
```diff
@@ -90,16 +87,16 @@ A Chart is composed of the following pieces:
 4. An optional `LICENSE` file
 5. An optional `docs/` directory
 6. An optional `hooks/` directory
```
Given that we don't currently implement any `hooks` functionality, I'd rather we drop this directory for now, and then add it back if and when we actually need it.

jackgr commented on docs/design/chart_format.md:
```diff
@@ -155,30 +149,23 @@ A definition file does not specify anything about individual pieces of the defin
+* version: A SemVer 2 semantic version of the chart (template files). Refer to the [instruction on semver.org](http://semver.org/).
+* keywords: A list of human-readable keywords. A keyword may contain alphanumeric text and spaces.
+* maintainers: A list of author objects, where an author object has two properties:
+   ** name: Author name
```
The double asterisk is not valid Markdown. It should be a single asterisk indented by 3 spaces.

jackgr commented on docs/design/chart_format.md:
```diff
@@ -155,30 +149,23 @@ A definition file does not specify anything about individual pieces of the defin
+* expander: Indicates how to process the contents of templates/
+   ** name: The name of the expander, as a Kubernetes service name or URL.
+   ** entrypoint: If the expander requires an entrypoint, gives the file (optional).
+* schema: The file used to validate the properties (user-configurable inputs) of this chart before expansion.  (optional)
```
Given that we are now standardizing on JSON schema v4, we should state that explicitly here.

sparkprime replied:
We haven't yet standardized on it, this is one of the items on #520.

jackgr commented on docs/design/chart_format.md:
```diff
@@ -189,25 +176,14 @@ keywords:
-The resolution of dependencies is not covered in this proposal, as it is an implementation detail of the DM service.
 ### Releasing a Chart
 A chart is _released_ when the source of the chart is tested, versioned, packaged into a gzipped tar file, and provenanced. Testing and provenancing attach badges to the Chart that attest to its quality and provenance. At that point, that particular release of a chart is considered immutable. No further changes are allowed. To enforce immutability through tamper detection, Charts must be digitall ...
 
```
Let's use `gs://kubernetes-charts` instead of `https://helm.sh/charts` in the example below, since we don't actually support the latter in the code.

jackgr commented on docs/design/chart_format.md:
```diff
@@ -234,7 +210,7 @@ $ helm release -r 1.1.2 ./nginx/
 A _Chart Repository_ is a place where _released copies_ of one or more charts may reside. (Unreleased charts are explicitly disallowed.) A Helm-DM Chart Repository is analogous to a Debian package repository or a Ruby Gems repository. It is a remote storage location accessible over a well-documented protocol (HTTP(S), and following fixed structural conventions.
```
We might want to relax the policy of disallowing unreleased charts. Either that, or we'll have to implement `helm chart release` for MVP and use it to release the charts that are currently sitting in `gs://kubernetes-charts` and friends.

jackgr replied:
What is Helm-DM? The official name of this product is Helm, right?

jackgr commented on docs/design/chart_format.md:
```diff
@@ -234,7 +210,7 @@ $ helm release -r 1.1.2 ./nginx/
+Chart repositories are styled after the predominant pattern of [object storage systems](https://cloud.google.com/storage/docs/key-terms). A _domain_ hosts one or more _repositories_. A repository is a bucket in which one or more charts may be stored. In an object storage system, this is represented by the pattern: **https://domain/bucket/object**. In object storage, the _object_ part of the URL m ...
 
 
 A chart name should be of the form _name-version.ext_, where _name_ is the chart name (alpha-numeric plus dash and underscore), and version is a SemVer v2 version. The extension, _ext_, should reflect the type of compression used to archive the object. This specification only discusses gzipped tar archives, but other mechanisms could be supported.
```
Regarding the following section, we currently don't support anything it describes. This content should either broken out into a "Future Directions" document or marked as aspirational and unsupported, so that readers can tell what's real and what isn't.

jackgr commented on docs/design/chart_format.md:
```diff
@@ -347,7 +323,7 @@ The format for a repository has packages stored in a flat namespace, with versio
 
 ### Aside: Why Not Git(Hub) or VCS?
-In their first incarnations, both Helm and DM used GitHub-backed repositories for released charts. The Helm team believes we made a huge mistake by going this route. This method suffers from some severe limitations (listed below). The proposal above is designed to work around those limitations.
+In their first incarnations, both Helm and DM used GitHub-backed repositories for released charts. The Helm team believes we made a huge mistake by going this route. This method suffers from some severe limitations (listed below). Our repository design does not have these limitations.
```
Second sentence... the Helm team. Reads like us vs. them. We were two teams, both of whom used Github. We are now one team and this design represents the thinking of the unified team. Please update the text accordingly.

jackgr commented on docs/design/chart_format.md:
```diff
@@ -394,7 +370,7 @@ The following URI schemes must be supported:
+The [Google Storage scheme](https://cloud.google.com/storage/docs/interoperability) (`gs://`)  and the S3 scheme (`s3://`) can also be supported. These forms do not follow the URL standard (bucket name is placed in the host field), but their behavior is well-documented.
 
 
 Examples:
```
In the Short Form section below, it should be noted that when a short form URL is used for an object storage service URL (e.g., starting with `gs` or `s3`), then the object storage service scheme appears as the `DOMAIN` segment of the reference (e.g., `helm:gs/kubernetes-charts/nginx-1.2.3.tgz`).

jackgr replied:
In Appendix A below, the Personas are not immediately relevant to this document. Suggest that we change them to:
* Chart Developer: Responsible for building new charts.
* App Developer: Responsible for building apps that use charts.

jackgr replied:
In the user stories below, typo `definied` should change to `defined`.

jackgr replied:
In the user stories below, some of the functionality described is speculative and not currently supported (e.g., search by various forms). Future looking stories should move to a design proposal.

jackgr replied:
In the user stories below, there's nothing that ties the proposed functionality to features of the product. For example, 

```
As a service dev, I want to be able to work with a team on the same definition
- As a service dev, I want to be able to work with a team on the same definition
- As a service dev, I want to be able to indicate when a particular definition is stable, and how stable it is
- As a service dev, I want to indicate the role I played in building a definition
```

We should not publish user stories without identifying how they are satisfied by the product, except in proposals.

jackgr replied:
In Appendix B below, we say that we've talked about having a standard (default) chart repository and we then suggest that it be Github based. This wording is problematic on two fronts:
* We already have a standard chart repository named `kubernetes-charts`. We should refer to it explicitly here.
* We have just told the reader why Github based repositories are not optimal. More importantly, we have stripped the Github repository support from the code.

sparkprime commented:
So I think I addressed all of these comments except the 

```
Chart Developer: Responsible for building new charts.
Application Developer: Responsible for building applications that use charts.
```

suggestion which I'm not confident is a good idea.  These are roles that we would be defining here, rather than roles people already understand.  I agree that section could use more work though.  It is now in a separate file.

jackgr commented:
What's wrong with Chart Developer and Application Developer?

People definitely understand Application Developer.

Chart Developer was a proposal to replace Definition Developer, which is totally vague.

sparkprime commented:
Ok I'll change

- Service Dev: Responsible for developing new charts

to Chart developer

sparkprime commented:
I'm not sure about application developer because we haven't defined what an application is, and the term is used in a lot of other contexts with very specific meanings.

jackgr commented:
OK. Then how about Chart User?

sparkprime commented:
Looks like I did in fact change it to "App dev" and just forgot about it, and then argued against doing what I'd already done.  Kinda shows you how much context switching I'm doing right now :)
Ready to LGTM?

jackgr commented:
LOL. Taking another look...

jackgr commented on docs/design/chart_format.md:
```diff
@@ -66,19 +59,18 @@ We assume that we are developing a technology that will enjoy widespread use amo
-Based on these, our design goal is to make it **easy to find, use, author, release, and maintain charts**, while still making it possible for advanced users to build sophisticated charts. Furthermore, we have begun the process of **attaching provenance to charts**.
+Based on these, our design goal is to make it **easy to find, use, author, release, and maintain charts**, while still making it possible for advanced users to build sophisticated charts.
 
+This design does not lose any of the functionality of Deployment Manager (DM), and which loses only Helm functionality that we believe is disposable. But both tools have changed substantial portions of their chart-handling logic.
```
This design is based on the integration of Deployment Manager (DM) and Helm, formerly at github.com/deis/helm. It does not lose any of the functionality of DM, and loses only Helm functionality that we believe is disposable. Substantial portions of the chart and template handling logic from the original implementations of these two tools have changed as a result of this integration. The new chart format, described in this document, is one of the primary drivers of those changes.

jackgr commented on docs/design/chart_format.md:
```diff
@@ -66,19 +59,18 @@ We assume that we are developing a technology that will enjoy widespread use amo
-
+This document does not define how either the client or the server must use this data, though we make some non-normative recommendations (e.g. “this data may be displayed to the user” or “a suitable backend will be able to translate this to a fully qualified URL”).
 
 Consequently, this document does not describe the implementation of either client or server sides of this spec. While it defines the pieces necessary for developing and deploying a chart, it does not define a development workflow. Some considerations are listed in [Appendix B](#appendix-b-the-standard-chart-repository).
```
Broken link. There is no Appendix B.

jackgr commented on docs/design/chart_format.md:
```diff
@@ -155,30 +137,23 @@ A definition file does not specify anything about individual pieces of the defin
+   * email: Author email (optional)
+* source: A URL to the source code for this chart
+* home: A URL to the home page for the project
+* expander: Indicates how to process the contents of templates/
```
This sub bullet is not displaying correctly in markdown. See the Markdown styling instructions at the link below. The problem is two asterisks. Should be one.

jackgr commented on docs/design/chart_format.md:
```diff
@@ -189,27 +164,16 @@ keywords:
-Dependencies hold between charts. That is, one chart may depend upon another chart.
-
+### Expanders and Templates
+The content of the `templates/` directory and the schema file are defined by the particular expander invoked by name in the Chart.yaml file.  Such expanders consume these files, in the context of properties, to generate content.  If a schema is given, the expander may use it to validate the properties before expansion.  Discussion of the available expanders and how they intepret the content of /t ...
```
Add...

If no expander is specified, files with `yaml` or `json is included without transformation in the chart output. Charts may therefore contain Kubernetes API objects

jackgr commented on docs/design/chart_format.md:
````diff
@@ -412,7 +304,13 @@ A generic short form maps to an object storage reference (`DOMAIN/REPOSITORY/REL
+helm:gs/kubernetes-charts/nginx-1.2.3.tgz
+```
+
+For the purpose of providing versioning ranges, and also for backward compatibility, version requirements may be treated as a suffix condition, instead of as part of the path:
````
Unless I'm mistaken, they MUST be treated as suffix conditions in short references. We do not recognize version information in the path (e.g., in the chart name or in an earlier path segment), and including it there will cause a parse error and/or a name matching failure.

jackgr commented on docs/design/chart_format.md:
```diff
@@ -447,41 +345,8 @@ Unprefixed relative paths are not valid. For example, `foo/` is not allowed as a
-## Appendix B: The Standard Chart Repository
-We have talked about adding a standard (default) chart repository that we are the gatekeepers on. This would hold only curated/vetted charts over which we have control. The following table illustrates how we’re thinking about implementing this:
+## Appendix A: The Standard Chart Repository
+The standard (default) chart repository is maintained by the Helm team at gs://kubernetes-charts/. This holds only curated/vetted charts over which we have control. The following table illustrates how we’re thinking about implementing this:
```
Actually, the standard (default) chart repository is maintained by the charts-maintainers team in the kubernetes organization. We expect these maintainers to diverge from the code maintainers over time, as people focused on content play a more active role.

jackgr replied:
@sparkprime Just noticed that this appendix is redundant with [team-workflows.md](https://github.com/kubernetes/helm/blob/master/docs/workflow/team-workflows.md). We should remove it from this document and update the team-workflows document to refer to the maintainers of `kubernetes/charts` per the comment above.

jackgr commented:
A few final nits. Otherwise, LGTM.
